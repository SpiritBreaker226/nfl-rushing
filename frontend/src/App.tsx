import React, { useState, useEffect } from 'react'

import axios from 'axios'

import { Player } from './types/Players'

import PlayersTable from './components/PlayersTable'

import './App.css'

function App() {
  const [isLoading, setIsLoading] = useState(true)
  const [errorMessage, setErrorMessage] = useState('')
  const [players, setPlayers] = useState<Player[]>([])
  const url = `${process.env.REACT_APP_BASE_API_URL}/players`

  useEffect(() => {
    const fetchDataFromApi = async () => {
      try {
        const res = await axios.get(url)
        // console.log('res', res.data.data)
        setPlayers(res.data.data)
        setIsLoading(false)
      } catch (error) {
        setErrorMessage(error.message)
        setIsLoading(false)
      }
    }

    fetchDataFromApi()
  }, [url, setIsLoading, setErrorMessage])

  if (isLoading) {
    return (
      <main className="App">
        <div className="loading">Loading...</div>
      </main>
    )
  }

  const Body = () => {
    if (errorMessage) {
      return <section className="error">{errorMessage}</section>
    }

    return (
      <section className="players">
        <PlayersTable players={players} />
      </section>
    )
  }

  return (
    <main className="App">
      <header className="App-header">
        <h1>Players</h1>
      </header>

      <Body />
    </main>
  )
}

export default App
