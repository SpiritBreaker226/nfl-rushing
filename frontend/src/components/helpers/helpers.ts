import { Dispatch } from 'react'

import queryString from 'query-string'

import { PlayerQuery } from '../../types/PlayerQuery'
import { Types } from '../../types/Actions'

export interface SetParamsCallbackProps {
  params: PlayerQuery
  dispatch: Dispatch<any>
  url: string
}

const setParamsCallback = ({
  params,
  dispatch,
  url,
}: SetParamsCallbackProps) => {
  const currentURL = queryString.parseUrl(url)
  const query = { ...currentURL.query, ...params }
  const newUrl = queryString.stringifyUrl(
    {
      url: currentURL.url,
      query,
    },
    {
      skipEmptyString: true,
      skipNull: true,
    }
  )

  dispatch({ type: Types.UpdateURL, payload: { url: newUrl } })
}

export { setParamsCallback }
