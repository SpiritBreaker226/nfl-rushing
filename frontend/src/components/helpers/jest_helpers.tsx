import React, { FunctionComponent } from 'react'

import { Player } from '../../types/Players'
import { Direction } from '../../types/PlayerQuery'
import { Pagination } from '../../types/Pagination'

import { AppContext } from '../../contexts/AppContext'
import { Action } from '../../types/Actions'

export const players: Player[] = [
  {
    attributes: {
      att: 2,
      attg: '2.0',
      avg: '3.5',
      first: 0,
      first_precentage: '0.0',
      fourty_plus: 0,
      fum: 0,
      lng: 7,
      player: 'Joe Banyard',
      pos: 'RB',
      td: 0,
      team: 'JAX',
      twenty_plus: 0,
      yds: 7,
      ydsg: '7.0',
    },
  },
  {
    attributes: {
      att: 205,
      attg: '12.8',
      avg: '5.1',
      first: 49,
      first_precentage: '23.9',
      fourty_plus: 2,
      fum: 2,
      lng: '75T',
      player: 'Mark Ingram',
      pos: 'RB',
      td: 6,
      team: 'NO',
      twenty_plus: 4,
      yds: 1,
      ydsg: '65.2',
    },
  },
  {
    attributes: {
      att: 3,
      attg: '0.8',
      avg: '-0.7',
      first: 0,
      first_precentage: '0.0',
      fourty_plus: 0,
      fum: 1,
      lng: 0,
      player: 'Brett Hundley',
      pos: 'QB',
      td: 0,
      team: 'GB',
      twenty_plus: 0,
      yds: -2,
      ydsg: '-0.5',
    },
  },
]

export interface TestingState {
  players?: Player[]
  pagination?: Pagination
  urlToPlayersEndpoint?: string
  search?: string
  errorMessage?: string
  isLoading?: boolean
  sorting?: {
    fieldName: string | undefined
    dir: Direction
  }
}

export interface MakeWrapperProps {
  state: TestingState
  dispatch?: (action: Action) => void
}

export const MakeWrapper: FunctionComponent<MakeWrapperProps> = ({
  state,
  dispatch,
  children,
}) => (
  <AppContext.Provider
    value={{
      state: {
        players: [],
        pagination: {
          rows_per_page: 10,
          page_total: 0,
          current_page: '1',
        },
        urlToPlayersEndpoint: '',
        search: '',
        errorMessage: '',
        sorting: {
          fieldName: undefined,
          dir: undefined,
        },
        isLoading: false,
        ...state,
      },
      dispatch: dispatch || (() => null),
    }}
  >
    {children}
  </AppContext.Provider>
)
