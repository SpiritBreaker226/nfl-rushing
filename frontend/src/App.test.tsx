import React from 'react'

import { waitFor, screen, render, fireEvent } from '@testing-library/react'

import axios from 'axios'

import { players } from './components/helpers/jest_helpers'

import App from './App'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('App', () => {
  describe('fetch player data', () => {
    beforeEach(async () => {
      mockedAxios.get.mockResolvedValueOnce({ data: { data: players } })

      render(<App />)

      await waitFor(() => screen.getByRole('heading'))
    })

    it('renders players component', () => {
      expect(mockedAxios.get).toHaveBeenCalledTimes(1)
      expect(mockedAxios.get).toHaveBeenCalledWith(
        `${process.env.REACT_APP_BASE_API_URL}/players`
      )
      expect(screen.getByRole('heading')).toHaveTextContent('Player')
      expect(screen.getByRole('main')).toMatchSnapshot()
    })

    describe('when searching', () => {
      it('find player name Brett', async () => {
        fireEvent.change(screen.getByRole('textbox'), {
          target: { value: 'brett' },
        })

        mockedAxios.get.mockResolvedValueOnce({ data: { data: [players[2]] } })

        fireEvent.click(screen.getByText('Search'))

        expect(
          screen.getByText('Loading', { exact: false })
        ).toBeInTheDocument()

        await waitFor(() => screen.getByRole('textbox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players?name=brett`
        )

        expect(screen.getByRole('textbox')).toHaveDisplayValue('brett')

        expect(screen.getByTestId('Brett HundleyGB')).toHaveTextContent(
          /brett/i
        )
        expect(screen.getAllByRole('row').length).toEqual(2)

        fireEvent.change(screen.getByRole('textbox'), {
          target: { value: 'mike' },
        })

        expect(screen.getByRole('textbox')).toHaveDisplayValue('mike')
      })

      it("reset player's table and URL on blank textbox", async () => {
        fireEvent.click(screen.getByText('Search'))

        await waitFor(() => screen.getByRole('heading'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players`
        )
        expect(screen.getByTestId('Mark IngramNO')).toHaveTextContent(/mark/i)
        expect(screen.getByTestId('Brett HundleyGB')).toHaveTextContent(
          /brett/i
        )
        expect(screen.getAllByRole('row').length).toEqual(4)
      })
    })

    describe('when download csv', () => {
      describe('url for the server should match what table looks like', () => {
        it('download a csv all of the players', async () => {
          expect(
            screen.getByText('Download', { exact: false }).getAttribute('href')
          ).toEqual(`${process.env.REACT_APP_BASE_API_URL}/players.csv`)
        })

        it('download a csv a player name mark', async () => {
          fireEvent.change(screen.getByRole('textbox'), {
            target: { value: 'mark' },
          })

          mockedAxios.get.mockResolvedValueOnce({
            data: { data: [players[2]] },
          })

          fireEvent.click(screen.getByText('Search'))

          expect(
            screen.getByText('Loading', { exact: false })
          ).toBeInTheDocument()

          await waitFor(() => screen.getByRole('textbox'))

          expect(
            screen.getByText('Download', { exact: false }).getAttribute('href')
          ).toEqual(
            `${process.env.REACT_APP_BASE_API_URL}/players.csv?name=mark`
          )
        })

        it('download a csv on sorted players', async () => {
          fireEvent.click(screen.getByTestId('yds'))

          await waitFor(() => screen.getByRole('textbox'))

          expect(
            screen.getByText('Download', { exact: false }).getAttribute('href')
          ).toEqual(
            `${process.env.REACT_APP_BASE_API_URL}/players.csv?sort_by=yds&sort_by_dir=asc`
          )
        })
      })
    })

    describe('when sorting', () => {
      it('should have the url to the server contain sort by and direction', async () => {
        fireEvent.click(screen.getByTestId('yds'))

        expect(
          screen.queryByText('Loading', { exact: false })
        ).toBeInTheDocument()

        await waitFor(() => screen.getByRole('textbox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players?sort_by=yds&sort_by_dir=asc`
        )

        fireEvent.click(screen.getByTestId('yds'))

        await waitFor(() => screen.getByRole('textbox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players?sort_by=yds&sort_by_dir=desc`
        )
      })

      it('should reset direction when changing sorted field', async () => {
        fireEvent.click(screen.getByTestId('yds'))

        await waitFor(() => screen.getByRole('textbox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players?sort_by=yds&sort_by_dir=asc`
        )

        fireEvent.click(screen.getByTestId('td'))

        await waitFor(() => screen.getByRole('textbox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players?sort_by=td&sort_by_dir=asc`
        )

        fireEvent.click(screen.getByTestId('td'))

        await waitFor(() => screen.getByRole('textbox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/players?sort_by=td&sort_by_dir=desc`
        )
      })
    })
  })

  describe('when loading the page', () => {
    it('should not render no player message and display loading text', async () => {
      mockedAxios.get.mockResolvedValueOnce({ data: { data: [] } })

      render(<App />)

      expect(
        screen.queryByText('Loading', { exact: false })
      ).toBeInTheDocument()

      expect(
        screen.queryByText('Download', { exact: false })
      ).not.toBeInTheDocument()

      expect(
        screen.queryByText('No Players Found', { exact: false })
      ).not.toBeInTheDocument()

      await waitFor(() => screen.getByRole('textbox'))

      expect(
        screen.queryByText('Loading', { exact: false })
      ).not.toBeInTheDocument()

      expect(
        screen.queryByText('Download', { exact: false })
      ).toBeInTheDocument()

      expect(
        screen.queryByText('No Players Found', { exact: false })
      ).toBeInTheDocument()
    })
  })
})
