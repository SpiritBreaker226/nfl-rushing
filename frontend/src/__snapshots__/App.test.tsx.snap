// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App fetch player data renders players component 1`] = `
<main>
  <section>
    <header
      class="makeStyles-header-3"
    >
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
        >
          <h1
            class="MuiTypography-root MuiTypography-h1"
          >
            Players
          </h1>
        </div>
        <div
          class="MuiGrid-root makeStyles-headerActions-2 MuiGrid-item MuiGrid-grid-xs-6"
        >
          <section>
            <div
              class="MuiFormControl-root MuiTextField-root makeStyles-searchText-4"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="searchBox"
                  name="searchBox"
                  placeholder="Search by Name"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Search
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </section>
        </div>
        <div
          class="MuiGrid-root makeStyles-headerActions-2 MuiGrid-item MuiGrid-grid-xs-4"
        >
          <section
            class="makeStyles-download-5"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-link-6 MuiTypography-colorPrimary"
              data-testid="download_csv"
              href="http://localhost:3000/players.csv"
            >
              Download CSV
            </a>
          </section>
        </div>
      </div>
    </header>
    <div
      class="makeStyles-appBody-1"
    >
      <section>
        <div
          class="MuiTableContainer-root"
        >
          <table
            class="MuiTable-root"
          >
            <thead
              class="MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Player's Name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Player's team abbreviation
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Player's postion
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing Attempts
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing Attempts Per Game Avg
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiTableSortLabel-root"
                    data-testid="yds"
                    role="button"
                    tabindex="0"
                  >
                    Total Rushing Yards
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing Avg Yards Per Attempt
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing Yards Per Game
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiTableSortLabel-root"
                    data-testid="td"
                    role="button"
                    tabindex="0"
                  >
                    Total Rushing Touchdowns
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiTableSortLabel-root"
                    data-testid="sortByLng"
                    role="button"
                    tabindex="0"
                  >
                    Longest Rush
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
                      />
                    </svg>
                  </span>
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing First Downs
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing First Down %
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing 20+ Yards Each
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing 40+ Yards Each
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head"
                  role="columnheader"
                  scope="col"
                >
                  Rushing Fumbles
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="2JAX"
                >
                  2
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="2.0JAX"
                >
                  2.0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="3.5JAX"
                >
                  3.5
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0JAX"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0.0JAX"
                >
                  0.0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0JAX"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0JAX"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="7JAX"
                >
                  7
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="Joe BanyardJAX"
                >
                  Joe Banyard
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="RBJAX"
                >
                  RB
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0JAX"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="JAXJAX"
                >
                  JAX
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0JAX"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="7JAX"
                >
                  7
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="7.0JAX"
                >
                  7.0
                </td>
              </tr>
              <tr
                class="MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="205NO"
                >
                  205
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="12.8NO"
                >
                  12.8
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="5.1NO"
                >
                  5.1
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="49NO"
                >
                  49
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="23.9NO"
                >
                  23.9
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="2NO"
                >
                  2
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="2NO"
                >
                  2
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="75TNO"
                >
                  75T
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="Mark IngramNO"
                >
                  Mark Ingram
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="RBNO"
                >
                  RB
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="6NO"
                >
                  6
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="NONO"
                >
                  NO
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="4NO"
                >
                  4
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="1NO"
                >
                  1
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="65.2NO"
                >
                  65.2
                </td>
              </tr>
              <tr
                class="MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="3GB"
                >
                  3
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0.8GB"
                >
                  0.8
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="-0.7GB"
                >
                  -0.7
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0GB"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0.0GB"
                >
                  0.0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0GB"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="1GB"
                >
                  1
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0GB"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="Brett HundleyGB"
                >
                  Brett Hundley
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="QBGB"
                >
                  QB
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0GB"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="GBGB"
                >
                  GB
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="0GB"
                >
                  0
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="-2GB"
                >
                  -2
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body"
                  data-testid="-0.5GB"
                >
                  -0.5
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <nav
          aria-label="pagination navigation"
          class="MuiPagination-root makeStyles-pagination-8"
          data-testid="pagination"
        >
          <ul
            class="MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiPaginationItem-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-current="true"
                aria-label="page 1"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded Mui-selected"
                tabindex="0"
                type="button"
              >
                1
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <button
                aria-label="Go to page 2"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
                tabindex="0"
                type="button"
              >
                2
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <button
                aria-label="Go to page 3"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
                tabindex="0"
                type="button"
              >
                3
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <button
                aria-label="Go to page 4"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
                tabindex="0"
                type="button"
              >
                4
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <button
                aria-label="Go to page 5"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
                tabindex="0"
                type="button"
              >
                5
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <div
                class="MuiPaginationItem-root MuiPaginationItem-ellipsis"
              >
                …
              </div>
            </li>
            <li>
              <button
                aria-label="Go to page 20"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
                tabindex="0"
                type="button"
              >
                20
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiPaginationItem-icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </li>
          </ul>
        </nav>
      </section>
    </div>
  </section>
</main>
`;
